hooks/tk-multi-publish2/nuke/publish_files.py

```diff
@@ -0,0 +1,263 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import nuke
+import sgtk
+import itertools
+from dd.runtime import api
+api.load("frangetools")
+import frangetools
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+
+class NukePublishFilesDDValidationPlugin(HookBaseClass):
+    """
+    Inherits from NukePublishFilesPlugin
+    """
+    def _build_dict(self, seq, key):
+        """
+        Creating a dictionary based on a key.
+
+        :param seq: list of dictionaries
+        :param key: dictionary key from which to create the dictionary
+        :return: dict with information for that particular key
+        """
+        return dict((d[key], dict(d, index=index)) for (index, d) in enumerate(seq))
+
+
+    def _framerange_to_be_published(self, item):
+        """
+        Since users have the option to render only a subset of frames,
+        adding validation to check if the full frame range is being published.
+
+        :param item: Item to process
+        :return: True if yes false otherwise
+        """
+        lss_path = item.properties['node']['cached_path'].value()
+        lss_data = frangetools.getSequence(lss_path)
+
+        # Since lss_data will be a list of dictionaries,
+        # building a dictionary from key value for the ease of fetching data.
+        info_by_path = self._build_dict(lss_data, key="path")
+        missing_frames = info_by_path.get(lss_path)['missing_frames']
+        root = nuke.Root()
+
+        # If there are no missing frames, then checking if the first and last frames match with root first and last
+        # Checking with root because _sync_frame_range() will ensure root is up to date with shotgun
+        if missing_frames:
+            self.logger.error("Incomplete renders! All the frames are not rendered.")
+            return False
+        else:
+            first_rendered_frame = info_by_path.get(lss_path)['frame_range'][0]
+            last_rendered_frame = info_by_path.get(lss_path)['frame_range'][1]
+            if (first_rendered_frame != root.firstFrame()) or (last_rendered_frame != root.lastFrame()):
+                self.logger.error("Incomplete renders! All the frames are not rendered.")
+                return False
+            return True
+
+
+    def _sync_frame_range(self, item):
+        """
+        Checks whether frame range is in sync with shotgun.
+
+        :param item: Item to process
+        :return: True if yes false otherwise
+        """
+        context = item.context
+        entity = context.entity
```

Comment:
You're right...context will always be valid since the context property will traverse up the item hierarchy if not defined locally, however it is quite possible for entity to be invalid, i.e. if this is a Project-level item, and it is also possible for the entity to not have sg_head_in/sg_tail_out attributes in its schema (i.e. its a Sequence or Asset).